third_party/libusb/naclport/src/chrome_usb/chrome-login-state-hook.js

```diff
@@ -0,0 +1,171 @@
+ * @extends goog.Disposable
+ * @constructor
+ */
+GSC.Libusb.ChromeLoginStateHook = function() {
+  /** 
+   * @type {boolean}
+   * @private  
+   */
+  this.simulateDevicesAbsent_ = false;
+  /** 
+   * @type {!Function}
+   * @private 
+   */
+  this.boundOnGotSessionState_ = this.onGotSessionState_.bind(this);
+  /** 
+   * @type {!goog.promise.Resolver}
+   * @private 
+   */
+  this.hookIsReadyResolver_ = goog.Promise.withResolver();
+  if (chrome.loginState) {
+    chrome.loginState.getProfileType(this.onGotProfileType_.bind(this));
+  } else {
+    GSC.Logging.getScopedLogger('Libusb.LoginStateHook').warning(
+        'chrome.loginState API is not available. This app might require a '
+        + 'newer version of Chrome.');
+    this.hookIsReadyResolver_.reject();
+  }
+};
+
+/** @const */
+const ChromeLoginStateHook = GSC.Libusb.ChromeLoginStateHook;
+
+goog.inherits(ChromeLoginStateHook, goog.Disposable);
+
+/**
+ * @type {!goog.log.Logger}
+ * @const
+ */
+ChromeLoginStateHook.prototype.logger = GSC.Logging.getScopedLogger(
+  'Libusb.LoginStateHook');
+
+/** @override */
+ChromeLoginStateHook.prototype.disposeInternal = function() {
+  if (chrome.loginState) {
+    chrome.loginState.onSessionStateChanged.removeListener(
+        this.boundOnGotSessionState_);
+  }
+
+  this.logger.fine('Disposed');
+
+  ChromeLoginStateHook.base(this, 'disposeInternal');
+};
+
+/**
+ * @return {function(string, !Array): !Array}
+ * @public
+ */ 
+ChromeLoginStateHook.prototype.getRequestSuccessHook = function() {
+  return this.requestSuccessHook_.bind(this);
+};
+
+/**
+ * @return {!goog.Promise}
+ * @public
+ */ 
+ChromeLoginStateHook.prototype.getHookReadyPromise = function() {
+  return this.hookIsReadyResolver_.promise;
+};
+
+ /**
+ * @param {!chrome.loginState.ProfileType} profileType
+ * @private
+ */
+ChromeLoginStateHook.prototype.onGotProfileType_ = function(profileType) {
+  goog.asserts.assert(chrome.loginState);
+  if (profileType === chrome.loginState.ProfileType.USER_PROFILE) {
+    chrome.loginState.getSessionState(this.boundOnGotSessionState_);
+    chrome.loginState.onSessionStateChanged.addListener(
+        this.onSessionStateChanged_.bind(this));
+  } else {
+    this.hookIsReadyResolver_.resolve();
+  }
+};
+
+/**
+ * @param {!chrome.loginState.SessionState} sessionState
+ * @private
+ */
+ChromeLoginStateHook.prototype.onGotSessionState_ = function(
+    sessionState) {
+  goog.asserts.assert(chrome.loginState);
+  if (sessionState === chrome.loginState.SessionState.IN_LOCK_SCREEN
+      && !this.simulateDevicesAbsent_) {
+    this.logger.info("No longer showing USB devices.");
+    this.simulateDevicesAbsent_ = true;
+  } else if (sessionState !== chrome.loginState.SessionState.IN_LOCK_SCREEN
+             && this.simulateDevicesAbsent_) {
+    this.logger.info("Showing USB devices.");
+    this.simulateDevicesAbsent_ = false;
+  }
+  this.hookIsReadyResolver_.resolve();
```

Comment:
nit: Probably worth adding a comment that this line can be hit multiple times, but only the first execution will actually resolve the promise and the further ones will be ignored.